### PR TITLE
Adds additional documentation to README.md

### PR DESCRIPTION
**Pull request by dlurton**
Improves `README.md` quite a bit.  Now finished.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,106 +1,341 @@
...
+val nullElement: AnyElement = loadSingleElement("null")
+val typedNullElement: AnyElement = loadSingleElement("null.int")
+
+// Both throw IonElementConstraintException
+nullElement.longValue
+typedNullElement.longValue
+```
+
+Each of `*Value` properties has a corresponding `*ValueOrNull`, which converts Ion null values to a Kotlin `null`. These
+should be used when the application wishes to allow elements to be non-null.
```
Comment (dlurton) [inline README.md:189]: *to be null.

File: README.md [hunk 2]
```diff
@@ -1,106 +1,341 @@
...
+val typedNullElement: AnyElement = loadSingleElement("null.int")
+println(typedNullElement.longValueOrNull)
+```
+
+TODO: describe the benefit of this with regard to Kotlin's null safety.
+
+### `AnyElement`'s `as*[OrNull]()` Functions
+
+`ion-element` greatly reduces the need for down-casting, but as shown above it does not does not completely
+
```
Comment (dlurton) [inline README.md:203]: *TODO

**Review by dlurton — COMMENTED** [review]
"Are there any cases where a user should choose to use IonValue over IonElement?"

File: README.md [hunk 3]
```diff
@@ -1,106 +1,341 @@
...
+      of `IonElement.asAnyElement()` at the call-site.
+    - ... a specific type of Ion data should use the narrowest possible `IonElement` sub-class, for instance:
+        - A function argument requiring an Ion s-expression should be of type `SexpElement`.
+        - A function argument allowing either a list *or* an s-expression should be of type `ContainerElement`.
+- Any functions that may return...
+    - ... any type of Ion data should have the `AnyElement` return type to avoid forcing the caller to down cast to a
+      narrower sub-interface or `AnyElement`.
+    - ... more than one type of Ion data should have the narrowest possible return type, for instance:
+        - A function that will only ever return s-expressions should have the `SexpElement` return type.
+        - A function that will return lists or s-expressions should have the `ContainerElement` return type.
```
Comment (dlurton) [inline README.md:220]: *`SeqElement`

File: README.md [hunk 4]
```diff
@@ -1,106 +1,541 @@
...
+that is idiomatic to Kotlin and is also usable from Java. It is meant as an alternative to `IonValue` from
+[Ion-java](https://github.com/amzn/ion-java) but is not a complete implementation of Ion as it relies on `Ion-java`'s
 `IonReader` and `IonWriter` interfaces for reading and writing Ion data.
 
-### Status
+## Status
 
-The the library is capable of representing any Ion value, however the public API should be considered 
-unstable and in alpha status.
+This library is complete and mature, however the public API should be considered unstable.
```
[the remark this answers is not in the document]
Comment (dlurton) [reply README.md:12]: So... I'd look at completing issues #53, #41 at least before cutting a `v1.0.0` release.  Also perhaps should consider #46, and #47.

File: README.md [hunk 5]
```diff
@@ -1,106 +1,541 @@
...
+
+// Struct is positioned on line 1 at character 1.
+assertEquals(IonTextLocation(1, 1), structElem.metas.location)
+// some_field begins on line 2 at character 3.
+assertEquals(IonTextLocation(2, 3), structElem["some_field"].metas.location)
 ```
 
+Note that `.location` above is an extension property provided for convenience.
+
+Finally, this works for both Ion text and Ion binary files, however for binary files, this the meta includes a byte
```
Comment (tgregg) [inline README.md:488]: this the meta -> the meta

**Review by tgregg — DISMISSED** [review]
Great documentation

File: README.md [hunk 6]
```diff
@@ -1,106 +1,541 @@
...
+`AnyElement` is a special sub-interface of `IonElement` that can be used by consumers of this library to reduce the
+overhead of accessing JVM native versions of the values represented by an `IonElement`. This is accomplished with two
+categories of accessor functions that perform type checking before returning a down-casted `IonElement` instance or
+Kotlin native value. As shown below, `AnyElement.as*[OrNull]()` functions perform type checking, throw an exception in
+the event of a type mismatch, and down-cast to the narrowed `IonElement` sub-interface.  `AnyElement.*Value[OrNull]`
+properties are similar but return JVM native values directly.
+
+The `IonElement.asAnyElement()` interface exists to convert to `AnyElement`. Although in `IonElement` implementations
+this is implemented as a simple downcast (i.e. `this as AnyElement`), implementations of `IonElement` are free to
+implement this however they see fit.
```
[the remark this answers is not in the document]
Comment (popematt) [reply README.md:131]:
```suggestion
The `IonElement.asAnyElement()` interface exists to convert to `AnyElement`. Although the implementations
of `IonElement` included in `ion-element-kotlin` implement this as a simple downcast (i.e. `this as AnyElement`), other implementations of `IonElement` are free to
implement this however they see fit.
```

File: README.md [hunk 7]
```diff
@@ -1,106 +1,541 @@
...
+}
+```
+
+More details about these functions are included below.
+
+## `AnyElement`'s `*Value[OrNull]` Properties
+
+In order to reduce the need to downcast to the appropriate sub-interface of `IonElement`, `AnyElement` provides two
+read-only properties per type of Ion data: `<kotlinType>Value` and `<kotlinType>ValueOrNull`, where `<jvmType>` is the
+JVM Kotlin equivalent of the corresponding Ion type. Each of these properties will check to if the Ion type is the
```
[the remark this answers is not in the document]
Comment (popematt) [reply README.md:174]:
```suggestion
JVM Kotlin equivalent of the corresponding Ion type. Each of these properties will check whether the Ion type is
```

File: README.md [hunk 8]
```diff
@@ -1,106 +1,541 @@
...
+```Kotlin
+val loader: IonElementLoader = createIonElementLoader(IonElementLoaderOptions(includeLocationMeta = true));
+val element: AnyElement = loader.loadSingleElement("{ some_field: 42 }")
+```
+
+## Writing Ion Data
+
+### `IonElement.toString()`
+
+The contract of `IonValue.toString()` does not guaranteed that the method will produce valid Ion text (though in practice
```
[the remark this answers is not in the document]
Comment (popematt) [reply README.md:426]:
```suggestion
The contract of `IonValue.toString()` does not guarantee that the method will produce valid Ion text (though in practice
```

File: README.md [hunk 9]
```diff
@@ -1,106 +1,541 @@
...
+it does produce valid Ion text). The constract of `IonElement.toString()` _does_ guarantee that the function must produce
```
[the remark this answers is not in the document]
Comment (popematt) [reply README.md:427]:
```suggestion
it does produce valid Ion text). The contract of `IonElement.toString()` _does_ guarantee that the function must produce
```

File: README.md (hunk 5 above)
Comment (popematt) [reply README.md:488]:
```suggestion
Finally, this works for both Ion text and Ion binary files, however for binary files, the meta includes a byte
```

**Review by popematt — APPROVED** [review]

**Pull request merged**